src/BoundaryConditions/fill_halo_regions_open.jl

```diff
@@ -1,39 +1,107 @@
-#####
-##### Outer functions for filling halo regions for open boundary conditions.
-#####
-
-# TODO: support true open boundary conditions.
-# For this we need to have separate functions for each of the six boundaries,
-# and need to unroll a loop over the boundary normal direction.
-# The syntax for `getbc` is also different for OpenBoundaryCondition than for others,
-# because the boundary-normal index can vary (and array boundary conditions need to be
-# 3D in general).
-
-@kernel function set_west_or_east_u!(u, i_boundary, bc, grid, args) 
+#=function fill_boundary_normal_velocities!(velocities, clock, fields)
+    grid = velocities.u.grid
+
+    fill_open_boundary_regions!(velocities.u, grid, instantiated_location(velocities.u), clock, fields)
+    fill_open_boundary_regions!(velocities.v, grid, instantiated_location(velocities.v), clock, fields)
+    fill_open_boundary_regions!(velocities.w, grid, instantiated_location(velocities.w), clock, fields)
+
+    return nothing
+end=#
+
+const BoundedGrid = Union{AbstractGrid{<:Any, <:Bounded}, 
+                          AbstractGrid{<:Any, <:Any, <:Bounded}, 
+                          AbstractGrid{<:Any, <:Any, <:Any, <:Bounded}}
+
+@inline fill_open_boundary_regions!(field, args...) = fill_open_boundary_regions!(field, field.grid, instantiated_location(field))
+
+@inline fill_open_boundary_regions!(field, grid::BoundedGrid, loc, args...) = fill_open_boundary_regions!(field, field.boundary_conditions, loc, grid, args...)
+
+function fill_open_boundary_regions!(field, boundary_conditions, loc, grid, args...; kwargs...)
+    arch = architecture(grid)
+
+    left_bc = left_boundary_condition(boundary_conditions, loc)
+    right_bc = right_boundary_condition(boundary_conditions, loc)
+
+    launch!(arch, grid, side_plan(loc), fill_open_halo!(loc), field, left_bc, right_bc, loc, grid, args)
+
+    return nothing
+end
+
+fill_open_boundary_regions!(fields::NTuple, boundary_conditions, loc, grid, args...; kwargs...) =
+    [fill_open_boundary_regions!(field, boundary_conditions[n], loc[n], grid, args...; kwargs...) for (n, field) in enumerate(fields)]
+
+@inline left_boundary_condition(boundary_conditions, loc) = nothing
+@inline left_boundary_condition(boundary_conditions, loc::Tuple{Face, Center, Center}) = boundary_conditions.west
+@inline left_boundary_condition(boundary_conditions, loc::Tuple{Center, Face, Center}) = boundary_conditions.south
+@inline left_boundary_condition(boundary_conditions, loc::Tuple{Center, Center, Face}) = boundary_conditions.bottom
+
+@inline right_boundary_condition(boundary_conditions, loc) = nothing
+@inline right_boundary_condition(boundary_conditions, loc::Tuple{Face, Center, Center}) = boundary_conditions.east
+@inline right_boundary_condition(boundary_conditions, loc::Tuple{Center, Face, Center}) = boundary_conditions.north
+@inline right_boundary_condition(boundary_conditions, loc::Tuple{Center, Center, Face}) = boundary_conditions.top
```

Comment:
shouldn't we have a method that accounts also for `Face, Face, Center`, `Face, Center, Face` and so on?
Or we just allow open in one direction?

Reply:
I'm not exactly sure what you mean here? We only need to do something different (i.e. do something in `fill_open_boundary_regions!`) for wall-normal velocity open boundary conditions to maintain the boundary adjacent cell to be divergence-free, but there is no reason todo anything else for any other open boundary condition (at least for the non-hydrostatic model, I'm not sure if we would need todo something different in the future for e.g. a free surface).

Greg and I worked through the maths for this, I'll add you to the overleaf and at some point add a summary of it to the top comment here.

Reply:
@jagoosw the problem is the name of this function. It's called "`left_boundary_condition`". But it returns `nothing` for locations other than the velocity locations apparently. When reading the code in isolation this doesn't make any sense. The other fields also have "left" and "right" boundary conditions. We need to find a more precise name for this function.

Reply:
Oh yes sorry this makes sense. I could change it to something more explicit like `left_face_boundary_condition` since that is actually what we want from it in this context?

Reply:
why is this needed at all? Can you explain the logic ?

Reply:
I'm not sure how else I could identify the boundary conditions on faces? If I just used `extract_bc` I would then need some way to get just the ones in the wall normal direction.

Reply:
Where are you using it

Reply:
Here:

https://github.com/CliMA/Oceananigans.jl/blob/92c16f33178d5fd6b352ff639d124be62920b7d0/src/BoundaryConditions/fill_halo_regions_open.jl#L19-L28

Where we just want to fill the open halos on face points.

Reply:
how about `left_open_boundary_condition`?

If you're only doing velocities then `left_velocity_open_boundary_condition` perhaps.